File: 0791-split-bst/0791-split-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def splitBST(self, root: Optional[TreeNode], target: int) -> List[Optional[TreeNode]]:
        # Base case: if root is None, return two None values
        if not root:
            return [None, None]
        
        # If root's value is greater than target, recursively split left subtree
        if root.val > target:
            left = self.splitBST(root.left, target)
            root.left = left[1]  # Attach the right part of the split to root's left subtree
            return [left[0], root]  # Return the left part and the root
        else:
            # Otherwise, recursively split right subtree
            right = self.splitBST(root.right, target)
            root.right = right[0]  # Attach the left part of the split to root's right subtree
            return [root, right[1]]  # Return the root and the right part

File: 0791-split-bst/0791-split-bst.cpp
class Solution {
public:
    vector<TreeNode*> splitBST(TreeNode* root, int target) {
        // Create dummy nodes to hold the smaller and larger parts
        TreeNode* dummySm = new TreeNode(0);
        TreeNode* curSm = dummySm;
        TreeNode* dummyLg = new TreeNode(0);
        TreeNode* curLg = dummyLg;

        // Start traversal from the root
        TreeNode* current = root;

        while (current != nullptr) {
            if (current->val <= target) {
                // Attach current node to the smaller part
                curSm->right = current;
                curSm = current;

                // Move to the right subtree
                TreeNode* nextNode = current->right;

                // Clear the right pointer of current node
                curSm->right = nullptr;

                current = nextNode;
            } else {
                // Attach current node to the larger part
                curLg->left = current;
                curLg = current;

                // Move to the left subtree
                TreeNode* nextNode = current->left;

                // Clear the left pointer of current node
                curLg->left = nullptr;

                current = nextNode;
            }
        }

        // Return the split parts as a vector
        return {dummySm->right, dummyLg->left};
    }
};
